internal/tool/ssh/impl/manager.go

```diff
@@ -448,57 +468,69 @@ func (m *manager) exportListener(_ context.Context, req *protos.ExportListenerRe
 }
 
 func (m *manager) startComponent(ctx context.Context, req *protos.ComponentToStart) error {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-
-	// Load app state.
-	state, _, err := m.loadAppState("" /*version*/)
-	if err != nil {
-		return err
+	// Record the component.
+	g := m.group(req.ColocationGroup)
+	record := func() bool {
+		g.components.Lock()
+		defer g.components.Unlock()
+		if g.components.Val[req.Component] {
+			// Component already started, or is in the process of being started.
+			return true
+		}
+		g.components.Val[req.Component] = true
+		return false
+	}
+	if record() { // already started
+		return nil
 	}
-	g := m.findOrAddGroup(state, req.ColocationGroup)
 
-	// Update routing information.
-	g.Components[req.Component] = req.IsRouted
+	// Update the routing info.
 	if req.IsRouted {
-		if _, ok := g.Assignments[req.Component]; !ok {
-			// Create an initial assignment for the component.
-			g.Assignments[req.Component] = &protos.Assignment{
+		update := func() {
+			g.routing.Lock()
+			defer g.routing.Unlock()
+			assignment := &protos.Assignment{
 				App:          m.dep.App.Name,
 				DeploymentId: m.dep.Id,
 				Component:    req.Component,
 			}
+			assignment = routingAlgo(assignment, g.routing.Val.Replicas)
+			g.routing.Val.Assignments = append(g.routing.Val.Assignments, assignment)
 		}
+		update()
 	}
-	if err := m.mayGenerateNewRoutingInfo(g); err != nil {
-		return err
-	}
-
-	// Store app state
-	m.appState.Update(appVersionStateKey, state)
 
-	// Start the colocation group, if it hasn't started already.
-	return m.startColocationGroup(ctx, &protos.ColocationGroup{Name: req.ColocationGroup})
+	// Start the colocation group, if it hasn't already started.
+	return m.startColocationGroup(g)
 }
 
-func (m *manager) startColocationGroup(_ context.Context, group *protos.ColocationGroup) error {
-	// If the group is already started, ignore.
-	if _, found := m.started[group.Name]; found {
+func (m *manager) startColocationGroup(g *group) error {
+	g.mu.Lock()
```

Comment:
maybe no need to hold the lock through starting below, i.e.,:

```
shouldStart := func() {
  g.mu.Lock()
  defer g.mu.Unlock()
  if g.started {
    return false
  }
  g.started = true
  return true
}
if !shouldStart() {
  return nil
}
```

It is a bit more verbose, so totallyh up to you.